apps/hannk/interpreter/model.h

```diff
@@ -349,30 +337,43 @@ class OpGroup : public Op {
         : Op(std::move(inputs), std::move(outputs)), ops_(std::move(ops)) {
     }
 
-    void add(OpPtr to_insert, const Op *before = nullptr);
-    void remove(const Op *op);
-
     BoundsMap map_bounds(int input_idx, int output_idx) const override;
 
     bool prepare() override;
-
     void execute() override;
 
     int op_count() const {
         return ops_.size();
     }
+    OpPtr op_ptr(int i) {
+        OpPtr result = nullptr;
+        std::swap(ops_[i], result);
+        return result;
+    }
+
+    // TODO: remove me
     Op *op(int i) {
         return ops_[i].get();
     }
     const Op *op(int i) const {
         return ops_[i].get();
     }
 
-    void accept(OpVisitor *v) override;
+    void accept(OpVisitor *v) const override;
+    OpPtr mutate(OpMutator *m, OpPtr op) override;
```

Comment:
I'll comment here for starters (and move into code if that's what you want)...

The problem here is that we are trying to implement double-dispatch via double-virtual-method calls (in the same fashion that Halide's similar classes do); the gotcha is that we *also* want to transfer ownership of 'this' into the receiving method call (since it will ultimately either keep-or-replace it). 

In Halide's IR, the `IntrusivePtr` approach allows us to transparently 'recover' a refcounted `Expr` from a specific pointer (e.g. `const Call *` implicitly converts to `Expr` and maintains the refcount correctly); with `unique_ptr`, there isn't any such way to do this, so we have to pass a redundant argument. It's not pretty, but the generated code is efficient.

For the record, I considered moving `Op` to use `shared_ptr` instead, which can sorta-kinda do the same sort of recovery via `std::enable_shared_from_this`, but it felt like it required too much obscure special knowledge to be friendly (e.g. explicit calls to the `shared_from_this()` method). I also considered moving `Op` to just use a cut-down variant of `IntrusivePtr`, but that would have been a pretty major chunk of code churn.

Reply:
I guess more specifically, I just think there should be a brief comment here on how to actually use this function: what the arguments mean.

Reply:
Added some comments, but also overhauled the mutate()/accept()/visit() methods some to be cleaner and to make improper calls hard-to-impossible, PTAL